chain/index/api.go

```diff
@@ -0,0 +1,345 @@
+		indexedData, err := si.getIndexedTipSetData(ctx, expectedTs)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to get indexed data for tipset at height %d: %w", expectedTs.Height(), err)
+		}
+		if indexedData == nil {
+			return nil, xerrors.Errorf("nil indexed data for tipset at height %d", expectedTs.Height())
+		}
+		if err = si.verifyIndexedData(ctx, expectedTs, indexedData); err != nil {
+			return nil, err
+		}
+		return indexedData, nil
+	}
+
+	indexedData, err := getAndVerifyIndexedData()
+	var bf bool
+	if err != nil {
+		if !backfill {
+			return nil, xerrors.Errorf("failed to verify indexed data at height %d: %w", expectedTs.Height(), err)
+		}
+
+		log.Warnf("failed to verify indexed data at height %d; err:%s; backfilling once and validating again", expectedTs.Height(), err)
+		if _, err := si.backfillMissingTipset(ctx, expectedTs); err != nil {
+			return nil, xerrors.Errorf("failed to backfill missing tipset at height %d during validation; err: %w", expectedTs.Height(), err)
+		}
+
+		indexedData, err = getAndVerifyIndexedData()
+		if err != nil {
+			return nil, xerrors.Errorf("failed to verify indexed data at height %d after backfill: %w", expectedTs.Height(), err)
+		}
+		bf = true
+	}
+
+	return &types.IndexValidation{
+		TipSetKey:                expectedTs.Key(),
+		Height:                   expectedTs.Height(),
+		IndexedMessagesCount:     uint64(indexedData.nonRevertedMessageCount),
+		IndexedEventsCount:       uint64(indexedData.nonRevertedEventCount),
+		IndexedEventEntriesCount: uint64(indexedData.nonRevertedEventEntriesCount),
+		Backfilled:               bf,
+	}, nil
+}
+
+func (si *SqliteIndexer) validateIsNullRound(ctx context.Context, epoch abi.ChainEpoch) (*types.IndexValidation, error) {
+	// make sure we do not have tipset(reverted or non-reverted) indexed at this epoch
+	var isNullRound bool
+	err := si.stmts.hasNullRoundAtHeightStmt.QueryRowContext(ctx, epoch).Scan(&isNullRound)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to check if null round exists at height %d: %w", epoch, err)
+	}
+	if !isNullRound {
+		return nil, xerrors.Errorf("index corruption: height %d should be a null round but is not", epoch)
+	}
+
+	return &types.IndexValidation{
+		Height:      epoch,
+		IsNullRound: true,
+	}, nil
+}
+
+func (si *SqliteIndexer) getTipsetCountsAtHeight(ctx context.Context, height abi.ChainEpoch) (revertedCount, nonRevertedCount int, err error) {
+	err = si.stmts.countTipsetsAtHeightStmt.QueryRowContext(ctx, height).Scan(&revertedCount, &nonRevertedCount)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			// No tipsets found at this height
+			return 0, 0, nil
+		}
+		return 0, 0, xerrors.Errorf("failed to query tipset counts at height %d: %w", height, err)
+	}
+
+	return revertedCount, nonRevertedCount, nil
+}
+
+type indexedTipSetData struct {
+	nonRevertedMessageCount      int
+	nonRevertedEventCount        int
+	nonRevertedEventEntriesCount int
```

Comment:
These could be `uint64`'s right? `Row#Scan()` should be fine with that I think and we don't have any query that would return a negative number. Then you wouldn't have to cast later when you use them.

Reply:
This is done.